apps/voting-disputable/contracts/DisputableVoting.sol

```diff
@@ -0,0 +1,745 @@
+    */
+    function _onDisputableActionAllowed(uint256 _voteId) internal {
+        Vote storage vote_ = votes[_voteId];
+        require(_isPaused(vote_), ERROR_VOTE_NOT_PAUSED);
+
+        vote_.disputableStatus = DisputableStatus.Active;
+        vote_.pauseDuration = getTimestamp64().sub(vote_.pausedAt);
+        emit ResumeVote(_voteId);
+    }
+
+    /**
+    * @dev Reject a vote
+    * @param _voteId Identification number of the vote to be rejected
+    */
+    function _onDisputableActionRejected(uint256 _voteId) internal {
+        Vote storage vote_ = votes[_voteId];
+        require(_isPaused(vote_), ERROR_VOTE_NOT_PAUSED);
+
+        vote_.disputableStatus = DisputableStatus.Cancelled;
+        vote_.pauseDuration = getTimestamp64().sub(vote_.pausedAt);
+        emit CancelVote(_voteId);
+    }
+
+    /**
+    * @dev Void an entry
+    * @param _voteId Identification number of the entry to be voided
+    */
+    function _onDisputableActionVoided(uint256 _voteId) internal {
+        _onDisputableActionAllowed(_voteId);
+    }
+
+    /**
+    * @dev Internal function to check if a vote can be executed. It assumes the queried vote exists.
+    * @return True if the given vote can be executed, false otherwise
+    */
+    function _canExecute(Vote storage vote_) internal view returns (bool) {
+        // If vote is already executed, it cannot be executed again
+        if (vote_.executed) {
+            return false;
+        }
+
+        // If the vote is paused or cancelled, it cannot be executed
+        if (_isPausedOrCancelled(vote_)) {
+            return false;
+        }
+
+        // If the vote is still open, it cannot be executed
+        if (_isVoteOpen(vote_)) {
+            return false;
+        }
+
+        // If the vote does not have enough support, it cannot be executed
+        uint256 totalVotes = vote_.yea.add(vote_.nay);
+        if (!_isValuePct(vote_.yea, totalVotes, vote_.supportRequiredPct)) {
+            return false;
+        }
+
+        // If the vote has not reached min quorum, it cannot be executed
+        if (!_isValuePct(vote_.yea, vote_.votingPower, vote_.minAcceptQuorumPct)) {
+            return false;
+        }
+
+        // If the vote cannot proceed due to an Agreement dispute, it cannot be executed
+        if (!_canProceedAgreementAction(vote_.actionId)) {
+            return false;
+        }
+
+        // If non of the above conditions are met, it can be executed
+        return true;
+    }
+
+    /**
+    * @dev Internal function to check if a voter can participate on a vote. It assumes the queried vote exists.
+    * @return True if the given voter can participate a certain vote, false otherwise
+    */
+    function _canVote(Vote storage vote_, address _voter) internal view returns (bool) {
+        return _isVoteOpen(vote_) && _canProceedAgreementAction(vote_.actionId) && token.balanceOfAt(_voter, vote_.snapshotBlock) > 0;
```

Comment:
The `_canProceedAgreementAction()` action may be a bit too expensive to do on every vote and I wonder if we can just check against the local state, e.g. with `_isVoteOpen()`.

Reply:
I think it's a matter of whether we want to fully trust the cache or not. This case, compared to executing the vote is different since it is not calling the Agreement (the execute calls `closeAgreementAction`), so we may be allowing someone to vote when it is actually not possible. We can decide to fully trust the cache tho.